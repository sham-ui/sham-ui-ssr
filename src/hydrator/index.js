import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,
        new ServerSideHooks( storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
