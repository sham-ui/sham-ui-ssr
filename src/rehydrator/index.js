import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
