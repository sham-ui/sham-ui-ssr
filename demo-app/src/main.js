import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

