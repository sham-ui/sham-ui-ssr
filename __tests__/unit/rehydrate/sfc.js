import { compile, compileAsSFC } from 'sham-ui-test-helpers';
import { ssrAndRehydrate } from '../../../src/testing';

it( 'should single file component work', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compileAsSFC`
        <template>
            {% if loaded %}
                Loaded!
            {% endif %}
        </template>
        
        <script>
            export default Component( Template );
        </script>
        `,
        {
            loaded: false
        }
    );
    expect( meta.html ).toBe( '<!--0-->' );
    meta.component.update( { loaded: true } );
    expect( meta.ctx.container.innerHTML ).toBe( ' Loaded! <!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should single file component correct work with options', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compileAsSFC`
        <template>
            {% if loaded %}
                {{text}}
            {% endif %}
        </template>
        
        <script>
            export default Component( Template, function( options ) {
                const text = $(); 
                options( {
                    [ text ]: {
                        get() {
                            return 'Text for content'
                        }
                    }
                } );
            } );
        </script>
        `,
        {
            loaded: false
        }
    );
    expect( meta.html ).toBe( '<!--0-->' );
    meta.component.update( { loaded: true } );
    expect( meta.ctx.container.innerHTML ).toBe( 'Text for content<!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should single file component correct work with context in blocks', async() => {
    expect.assertions( 2 );
    window.CustomPanel = compile`
        <div>
            <div class="title">
                {% defblock 'title' %}
            </div>
        </div>
    `;
    const meta = await ssrAndRehydrate(
        compileAsSFC`
        <template>
            <CustomPanel>
                {% block 'title' %}
                    {{this.title()}}
                {% endblock %}
            </CustomPanel>
        </template>
        
        <script>
            export default Component( Template, function() {
                this.title = () => {
                    return 'Title text'
                };
            } );
        </script>
        `
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><div class="title"> <!--0-->Title text<!--1--> <!--0--></div></div><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.CustomPanel;
} );

it( 'should work with class getters in expressions', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compileAsSFC`
        <template>
            <span>{{this.user}}</span>
        </template>
        
        <script>
            export default Component( Template, function() {
                this.user = 'John Smith'; 
            } );
        </script>
        `
    );
    expect( meta.html ).toBe( '<span>John Smith</span>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with class getters in if', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compileAsSFC`
        <template>
            {% if this.isVisible %}
                <span>{{user}}</span>
            {% endif %}
        </template>
        
        <script>
            export default Component( Template, function() {
                this.isVisible = true;
            } );
        </script>
        `,
        {
            user: 'Joh Smith'
        }
    );
    expect( meta.html ).toBe( '<span>Joh Smith</span><!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with class property in for', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compileAsSFC`
        <template>
            <ul>
                {% for user of userList %}
                    <li>{{user}}</li>
                {% endfor %}
            </ul>
        </template>
        
        <script>
            export default Component( Template, function( options ) {
                const userList = $(); 
                options( {
                    [ userList ]: [ 
                        'John Smith',
                        'Adam Mock'
                     ]
                } );
            } );
        </script>
        `
    );
    expect( meta.html ).toBe(
        '<ul><li>John Smith</li><li>Adam Mock</li></ul>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
