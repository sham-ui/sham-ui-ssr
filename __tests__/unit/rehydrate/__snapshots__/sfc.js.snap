// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should single file component correct work with context in blocks 1`] = `
Object {
  "Options": Object {},
  "html": "
<div>
  <div class=\\"title\\">
    <!--0-->Title text
    <!--1-->
    <!--0-->
  </div>
</div>
<!--0-->
",
}
`;

exports[`should single file component correct work with options 1`] = `
Object {
  "Options": Object {
    "loaded": true,
    "text": "Text for content",
  },
  "html": "
Text for content
<!--0-->
",
}
`;

exports[`should single file component work 1`] = `
Object {
  "Options": Object {
    "loaded": true,
  },
  "html": "
 Loaded!
 <!--0-->
",
}
`;

exports[`should work with class getters in expressions 1`] = `
Object {
  "Options": Object {},
  "html": "<span>John Smith</span>",
}
`;

exports[`should work with class getters in if 1`] = `
Object {
  "Options": Object {
    "user": "Joh Smith",
  },
  "html": "
<span>Joh Smith</span>
<!--0-->
",
}
`;

exports[`should work with class property in for 1`] = `
Object {
  "Options": Object {
    "userList": Array [
      "John Smith",
      "Adam Mock",
    ],
  },
  "html": "
<ul>
  <li>John Smith</li>
  <li>Adam Mock</li>
</ul>
",
}
`;
